File: android/src/main/java/com/visioncameramlkit/bridge/plugins/BaseMLKitPlugin.kt
package com.visioncameramlkit.bridge.plugins

import android.util.Log
import com.google.mlkit.common.MlKitException
import com.mrousavy.camera.core.FrameInvalidError
import com.mrousavy.camera.frameprocessors.Frame
import com.mrousavy.camera.frameprocessors.FrameProcessorPlugin
import com.mrousavy.camera.frameprocessors.VisionCameraProxy

@Suppress("MagicNumber")
abstract class BaseMLKitPlugin(
  @Suppress("unused")
  protected val proxy: VisionCameraProxy,
  protected val options: Map<String, Any>?,
) : FrameProcessorPlugin() {
  protected abstract val tag: String

  private val frameProcessInterval: Int =
    (options?.get("frameProcessInterval") as? Number)?.toInt() ?: 0
  private var frameCounter: Int = 0
  protected val invertColors: Boolean = options?.get("invertColors") as? Boolean ?: false
  protected val scaleFactor: Float =
    (options?.get("scaleFactor") as? Number)?.toFloat()?.coerceIn(0.9f, 1.0f) ?: 1.0f

  // Android handles rotation automatically, so outputOrientation is ignored. Added for iOS parity.
  protected val outputOrientation: com.visioncameramlkit.domain.models.OutputOrientation? =
    com.visioncameramlkit.domain.models.OutputOrientation
      .fromString(options?.get("outputOrientation") as? String)

  protected abstract fun processFrame(
    frame: Frame,
    arguments: Map<String, Any>?,
  ): HashMap<String, Any>

  override fun callback(
    frame: Frame,
    arguments: Map<String, Any>?,
  ): HashMap<String, Any> {
    // Process 1 frame, skip N frames, repeat
    // Note: Consider using runAtTargetFps() from Vision Camera instead
    if (frameProcessInterval > 0) {
      frameCounter++
      if (frameCounter <= frameProcessInterval) {
        return hashMapOf()
      }
      frameCounter = 0
    }

    return try {
      processFrame(frame, arguments)
    } catch (e: FrameInvalidError) {
      Log.e(tag, "FrameInvalidError occurred while processing the frame", e)
      hashMapOf()
    } catch (e: MlKitException) {
      Log.e(tag, "MlKitException occurred while processing the image with ML Kit", e)
      hashMapOf()
    } catch (
      @Suppress("TooGenericExceptionCaught") e: Exception,
    ) {
      Log.e(tag, "Unexpected error occurred during recognition", e)
      hashMapOf()
    }
  }
}

File: ios/Bridge/Plugins/BaseMLKitPlugin.swift
import Foundation

#if canImport(VisionCamera)
  import VisionCamera

  @objc
  public class BaseMLKitPlugin: FrameProcessorPlugin {

    let proxy: VisionCameraProxyHolder
    let options: [AnyHashable: Any]?

    private let frameProcessInterval: Int
    private var frameCounter: Int = 0
    let invertColors: Bool
    let scaleFactor: CGFloat

    var outputOrientation: OutputOrientation?

    public override init(
      proxy: VisionCameraProxyHolder,
      options: [AnyHashable: Any]?
    ) {
      self.proxy = proxy
      self.options = options
      self.frameProcessInterval = options?["frameProcessInterval"] as? Int ?? 0
      self.invertColors = options?["invertColors"] as? Bool ?? false

      let sf = options?["scaleFactor"] as? NSNumber
      let sfValue = sf?.doubleValue ?? 1.0

      self.scaleFactor = min(max(CGFloat(sfValue), 0.9), 1.0)

      super.init(proxy: proxy, options: options)
    }

    public override func callback(
      _ frame: Frame,
      withArguments arguments: [AnyHashable: Any]?
    ) -> Any {
      // Extract outputOrientation from arguments for rotation support
      outputOrientation = OutputOrientation(
        string: arguments?["outputOrientation"] as? String
      )

      // Process 1 frame, skip N frames, repeat
      // Note: Consider using runAtTargetFps() from Vision Camera instead
      if frameProcessInterval > 0 {
        frameCounter += 1
        if frameCounter <= frameProcessInterval {
          return [:]
        }
        frameCounter = 0
      }

      do {
        return try processFrame(frame, withArguments: arguments)
      } catch {
        print("[\(pluginName)] Error occurred: \(error.localizedDescription)")
        return [:]
      }
    }

    public func processFrame(
      _ frame: Frame,
      withArguments arguments: [AnyHashable: Any]?
    ) throws -> [String: Any] {
      fatalError("processFrame must be overridden by subclass")
    }

    var pluginName: String {
      return String(describing: type(of: self))
    }
  }
#endif
